livekit-rtc/livekit/rtc/chat.py

```diff
@@ -0,0 +1,140 @@
+# Copyright 2023 LiveKit, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass, field
+from datetime import datetime
+import json
+import logging
+from typing import Any, Callable, Dict, Optional
+
+from .room import Room, Participant, DataPacket
+from ._proto.room_pb2 import DataPacketKind
+from ._utils import generate_random_base62
+
+_CHAT_TOPIC = "lk-chat-topic"
+_CHAT_UPDATE_TOPIC = "lk-chat-update-topic"
+
+
+class ChatManager:
+    """A utility class that sends and receives chat messages in the active session.
+
+    It implements LiveKit Chat Protocol, and serializes data to/from JSON data packets.
+    """
+
+    def __init__(
+        self, room: Room, *, on_message: Callable[["ChatMessage"], None] = None
+    ):
+        self._lp = room.local_participant
+        self._room = room
+        self._callback: Callable[["ChatMessage"], None] = None
+
+        room.on("data_received", self._on_data_received)
+        if on_message:
+            self.on_message(on_message)
+
+    def close(self):
+        self._room.off("data_received", self._on_data_received)
+
+    async def send_message(self, message: str) -> "ChatMessage":
+        """Send a chat message to the end user using LiveKit Chat Protocol.
+
+        Args:
+            message (str): the message to send
+
+        Returns:
+            ChatMessage: the message that was sent
+        """
+        msg = ChatMessage(
+            message=message,
+            is_local=True,
+            participant=self._lp,
+        )
+        await self._lp.publish_data(
+            payload=json.dumps(msg.asjsondict()),
+            kind=DataPacketKind.KIND_RELIABLE,
+            topic=_CHAT_TOPIC,
+        )
+        return msg
+
+    async def update_message(self, message: "ChatMessage"):
+        """Update a chat message that was previously sent.
+
+        If message.deleted is set to True, we'll signal to remote participants that the message
+        should be deleted.
+        """
+        await self._lp.publish_data(
+            payload=json.dumps(message.asjsondict()),
+            kind=DataPacketKind.KIND_RELIABLE,
+            topic=_CHAT_UPDATE_TOPIC,
+        )
+
+    def on_message(self, callback: Callable[["ChatMessage"], None]):
```

Comment:
We don't have on_message_updated?
IMO let's use https://github.com/livekit/python-sdks/blob/main/livekit-rtc/livekit/rtc/_event_emitter.py since we're already using it on other places

Reply:
will do